### PR TITLE
functional_tests: add test for private registry auth over asm

### PR DESCRIPTION
**Pull request by adnxn**
### Summary

https://github.com/adnxn/amazon-ecs-agent/commits/pra-ft
### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [ ] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [ ] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: 

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by yunhee-l** [comment]
is this not a supported feature in ECS Windows?

**Comment by adnxn** [comment]
@yunhee-l, it is but right now we don't have a local private registry set up windows like we have on linux

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 1]
```diff
@@ -839,6 +841,56 @@ func TestAWSLogsDriverDatetimeFormat(t *testing.T) {
 	assert.Equal(t, *resp.Events[1].Message, "May 01, 2017 19:00:04 Agent\nRunning\nin the instance\n", fmt.Sprintf("Got log events message unexpected: %s", *resp.Events[1].Message))
 }
 
+// This tests the workflow for retriving private registry authentication data
```
Comment (richardpen) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: nit: TestPrivateRegistryAuthOverASM tests xxx

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 2]
```diff
@@ -839,6 +841,56 @@ func TestAWSLogsDriverDatetimeFormat(t *testing.T) {
...
+// from AWS Secrets Manager
+func TestPrivateRegistryAuthOverASM(t *testing.T) {
+
+	if os.Getenv("TEST_DISABLE_EXECUTION_ROLE") == "true" {
+		t.Skip("TEST_DISABLE_EXECUTION_ROLE was set to true")
+	}
+
+	RequireDockerVersion(t, ">=17.06.0-ce")
```
Comment (richardpen) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: Why is this required for this test?

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 3]
```diff
@@ -839,6 +841,56 @@ func TestAWSLogsDriverDatetimeFormat(t *testing.T) {
...
+			default:
+				require.NoError(t, err, "AWS Secrets Manager CreateSecret call failed")
+			}
+		}
+	}
+
+	agent := RunAgent(t, nil)
+	defer agent.Cleanup()
+
+	agent.RequireVersion(">=1.18.0")
```
Comment (richardpen) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: Shouldn't this be 1.19.0?

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 4]
```diff
@@ -839,6 +841,56 @@ func TestAWSLogsDriverDatetimeFormat(t *testing.T) {
...
+
+	tdOverrides := make(map[string]string)
+	tdOverrides["$$$REPOSITORY_CREDENTIALS$$$"] = secretName
+	tdOverrides["$$$EXECUTION_ROLE$$$"] = os.Getenv("ECS_FTS_EXECUTION_ROLE")
+
+	task, err := agent.StartTaskWithTaskDefinitionOverrides(t, "private-registry-auth-asm-validator-unix", tdOverrides)
+	require.NoError(t, err, "Expected to start task using private registry authentication over asm failed")
+
+	err = task.WaitStopped(2 * time.Minute)
```
Comment (richardpen) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: Can you use the constants defined [here](https://github.com/aws/amazon-ecs-agent/blob/master/agent/functional_tests/tests/functionaltests_test.go#L42)

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 5]
```diff
@@ -839,6 +841,56 @@ func TestAWSLogsDriverDatetimeFormat(t *testing.T) {
...
+	require.NoError(t, err)
+	exitCode, _ := task.ContainerExitcode("private-registry-auth-asm-validator")
+	assert.Equal(t, 42, exitCode, fmt.Sprintf("Expected exit code of 42; got %d", exitCode))
+}
+
 // TestContainerHealthMetrics tests the container health metrics was sent to backend
```
Comment (richardpen) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: Since you are here, can you remove the line 896,897 and 904,905?

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 2 above)
Comment (adnxn) [reply agent/functional_tests/tests/functionaltests_unix_test.go]: 🤔 you're right, its not required

File: agent/functional_tests/tests/functionaltests_unix_test.go (hunk 3 above)
Comment (adnxn) [reply agent/functional_tests/tests/functionaltests_unix_test.go]: updating VERSION file too.

**Comment by adnxn** [comment]
@aws/aws-ecs-agent also updated VERSION file to `1.19.0-dev`

File: VERSION [hunk 6]
```diff
@@ -1 +1 @@
-1.18.0
\ No newline at end of file
+1.19.0-dev
```
Comment (yunhee-l) [inline VERSION]: I thought we only update version file during agent release?

Comment (adnxn) [reply VERSION]: we only remove the `-dev` bit during release, otherwise we keep this for new agent version dependent functional tests to not break CI
